Sources/Valkey/ValkeyClient.swift

```diff
@@ -209,3 +224,112 @@ extension ValkeyClient {
 @available(valkeySwift 1.0, *)
 extension ValkeyClient: Service {}
 #endif  // ServiceLifecycle
+
+@available(valkeySwift 1.0, *)
+@usableFromInline
+enum RequestState: AtomicRepresentable, Sendable {
+    @usableFromInline
+    typealias AtomicRepresentation = Unmanaged<ValkeyConnection>?
+
+    case waitingForConnection
+    case onConnection(ValkeyConnection)
+
+    @usableFromInline
+    static func decodeAtomicRepresentation(_ storage: consuming Unmanaged<ValkeyConnection>?) -> RequestState {
+        if let storage {
+            return .onConnection(storage.takeRetainedValue())
+        } else {
+            return .waitingForConnection
+        }
+    }
+
+    @usableFromInline
+    static func encodeAtomicRepresentation(_ value: consuming RequestState) -> Unmanaged<ValkeyConnectionRequest.Connection>? {
+        switch value {
+        case .onConnection(let connection):
+            return Unmanaged.passRetained(connection)
+        case .waitingForConnection:
+            return nil
+        }
+    }
+}
+
+@available(valkeySwift 1.0, *)
+@usableFromInline
+final class ValkeyConnectionRequest<T: Sendable>: Sendable, ConnectionRequestProtocol {
+    @usableFromInline
+    typealias Connection = ValkeyConnection
+
+    @usableFromInline
+    let id: Int
+    @usableFromInline
+    let pool: ValkeyClient.Pool
+    @usableFromInline
+    let continuation: CheckedContinuation<T, any Error>
+    @usableFromInline
+    let lock: Mutex<RequestState>
+    @usableFromInline
+    let onConnection: @Sendable (Connection, ValkeyConnectionRequest<T>) -> ()
+
+    @inlinable
+    init(
+        id: Int,
+        pool: ValkeyClient.Pool,
+        continuation: CheckedContinuation<RESPToken, any Error>,
+        _ onConnection: @escaping @Sendable (Connection, ValkeyConnectionRequest<T>) -> ()
+    ) where T == RESPToken {
+        self.id = id
+        self.pool = pool
+        self.continuation = continuation
+        self.onConnection = onConnection
+        self.lock = .init(.waitingForConnection)
+    }
+
+    @inlinable
+    func complete(with result: Result<Connection, ConnectionPoolError>) {
+        switch result {
+        case .success(let connection):
+            self.lock.withLock { state in
+                state = .onConnection(connection)
+            }
+            self.onConnection(connection, self)
```

Comment:
complete and succeed seem to do different things for a successful result, which is the correct one?